### ISSUE TITLE
diffie-hellman-group-exchange-sha256

### ISSUE THREAD

**Issue by perkelix** [jtesta/ssh-audit#203]
Version 2.9.0 seemingly has a self-contradictory perspective on `diffie-hellman-group-exchange-sha256`:

If it's included as a kex, performing the audit warns that...
```
(kex) diffie-hellman-group-exchange-sha256 (2048-bit) -- [warn] 2048-bit modulus only provides 112-bits of symmetric strength
                                                      `- [info] available since OpenSSH 4.4
                                                      `- [info] A bug in OpenSSH causes it to fall back to a 2048-bit modulus regardless of server configuration (https://bugzilla.mindrot.org/show_bug.cgi?id=2793)
```

If it's omitted, auditing (without command option) recommends adding it, while auditing by policy flat out results in a failed mark.

Please make up your mind: EITHER recommend OR discourage its inclusion. Thanks.

**Comment by thecliguy** [comment]
I concur, there does indeed seem to be a contradiction if the target server is running OpenSSH... 

The algorithm recommendations section of ssh-audit advises the use of `diffie-hellman-group-exchange-sha256`, however due to the following bug in OpenSSH, ssh-audit will always complain about its presence: `A bug in OpenSSH causes it to fall back to a 2048-bit modulus regardless of server configuration (https://bugzilla.mindrot.org/show_bug.cgi?id=2793)`.

@jtesta - Because of the aforementioned bug, should `diffie-hellman-group-exchange-sha256` be removed as a recommended algorithm if the target server is running OpenSSH?

**Comment by jtesta** [comment]
I just checked in a fix to further refine the results against OpenSSH (see https://github.com/jtesta/ssh-audit/commit/199e75f6cdf5676c9b4b08907db7d48f12f7fa72).  If the server's fallback mechanism is suspected of being triggered, an additional test will be run to obtain a more accurate result.

I think this addresses the core problem.  If not, @perkelix, please feel free to re-open the issue.  Thanks for reporting!

**Comment by perkelix** [comment]
That didn't solve the issue. Without the algorithm:

```
$ ./ssh-audit.py -P "Hardened Ubuntu Server 22.04 LTS (version 3)" 172.16.1.1
Host:   172.16.1.1
Policy: Hardened Ubuntu Server 22.04 LTS (version 3)
Result: ❌ Failed!

Errors:
  * Key exchanges did not match.
    - Expected: sntrup761x25519-sha512@openssh.com, curve25519-sha256, curve25519-sha256@libssh.org, diffie-hellman-group16-sha512, diffie-hellman-group18-sha512, diffie-hellman-group-exchange-sha256
    - Actual:   sntrup761x25519-sha512@openssh.com, curve25519-sha256, curve25519-sha256@libssh.org, diffie-hellman-group16-sha512, diffie-hellman-group18-sha512
```

```
# algorithm recommendations (for OpenSSH 9.2)
(rec) +diffie-hellman-group-exchange-sha256-- kex algorithm to append 
```

With the algorithm:

```
$ ./ssh-audit.py -P "Hardened Ubuntu Server 22.04 LTS (version 3)" 172.16.1.1
Host:   172.16.1.1
Policy: Hardened Ubuntu Server 22.04 LTS (version 3)
Result: ❌ Failed!

Errors:
  * Group exchange (diffie-hellman-group-exchange-sha256) modulus sizes did not match.
    - Expected: 4096
    - Actual:   3072
```

**Comment by perkelix** [comment]
@jtesta That didn't solve the issue. Without the algorithm:

```
$ ./ssh-audit.py -P "Hardened Ubuntu Server 22.04 LTS (version 3)" 172.16.1.1
Host:   172.16.1.1
Policy: Hardened Ubuntu Server 22.04 LTS (version 3)
Result: ❌ Failed!

Errors:
  * Key exchanges did not match.
    - Expected: sntrup761x25519-sha512@openssh.com, curve25519-sha256, curve25519-sha256@libssh.org, diffie-hellman-group16-sha512, diffie-hellman-group18-sha512, diffie-hellman-group-exchange-sha256
    - Actual:   sntrup761x25519-sha512@openssh.com, curve25519-sha256, curve25519-sha256@libssh.org, diffie-hellman-group16-sha512, diffie-hellman-group18-sha512
```

```
# algorithm recommendations (for OpenSSH 9.2)
(rec) +diffie-hellman-group-exchange-sha256-- kex algorithm to append 
```

With the algorithm:

```
$ ./ssh-audit.py -P "Hardened Ubuntu Server 22.04 LTS (version 3)" 172.16.1.1
Host:   172.16.1.1
Policy: Hardened Ubuntu Server 22.04 LTS (version 3)
Result: ❌ Failed!

Errors:
  * Group exchange (diffie-hellman-group-exchange-sha256) modulus sizes did not match.
    - Expected: 4096
    - Actual:   3072
```

**Comment by jtesta** [comment]
@perkelix : I investigated this, and found it is caused by a discrepancy between the policies shipped in v3.0.0 and the hardening guides on the website.  This was not found sooner because the OpenSSH fallback mechanism (and our prior method for handling it) masked the issue.  I've opened #205 to add functionality to address this condition.

In the meantime, to fix this on your end, you can run the following on your server to enable only 4096-bit DH moduli or larger:
```
awk '$5 >= 4095' /etc/ssh/moduli > /etc/ssh/moduli.safe
mv /etc/ssh/moduli.safe /etc/ssh/moduli
```

**Comment by perkelix** [comment]
Reopening.

> @perkelix : I investigated this, and found it is caused by a discrepancy between the policies shipped in v3.0.0 and the hardening guides on the website. This was not found sooner because the OpenSSH fallback mechanism (and our prior method for handling it) masked the issue. I've opened #205 to add functionality to address this condition.
> 
> In the meantime, to fix this on your end, you can run the following on your server to enable only 4096-bit DH moduli or larger:
> 
> ```
> awk '$5 >= 4095' /etc/ssh/moduli > /etc/ssh/moduli.safe
> mv /etc/ssh/moduli.safe /etc/ssh/moduli
> ```

That workaround (and bug #205's ideas of lowering the minimum to 3072 bits) further emphasizes how this bug isn't fixed. If an SSH server implementation is known to have a buggy implementation of a specific algorithm, just don't recommend including it.
